address-space-controller/src/main/java/io/enmasse/controller/auth/CertBundleCertProvider.java

```diff
@@ -28,34 +28,40 @@ public CertBundleCertProvider(OpenShiftClient client) {
-        if (secret == null) {
-            Map<String, String> data = new HashMap<>();
-            String tlsKey = endpointInfo.getCertSpec().getTlsKey();
-            String tlsCert = endpointInfo.getCertSpec().getTlsCert();
-            if (tlsKey == null) {
-                log.warn("tlsKey not present, not providing cert for {}", endpointInfo.getServiceName());
-                return;
-            }
-
-            if (tlsCert == null) {
-                log.warn("tlsCert not present, not providing cert for {}", endpointInfo.getServiceName());
-                return;
-            }
-
-            data.put("tls.key", tlsKey);
-            data.put("tls.crt", tlsCert);
-            log.info("Creating cert secret with certBundle input");
-            client.secrets().inNamespace(namespace).withName(endpointInfo.getCertSpec().getSecretName()).createNew()
-                    .editOrNewMetadata()
-                    .withName(endpointInfo.getCertSpec().getSecretName())
-                    .addToLabels(LabelKeys.INFRA_UUID, addressSpace.getAnnotation(AnnotationKeys.INFRA_UUID))
-                    .addToLabels(LabelKeys.INFRA_TYPE, addressSpace.getType())
-                    .addToLabels("app", "enmasse")
-                    .endMetadata()
-                    .withType("kubernetes.io/tls")
-                    .withData(data)
-                    .done();
+        Map<String, String> data = new HashMap<>();
+        String tlsKey = endpointInfo.getCertSpec().getTlsKey();
+        String tlsCert = endpointInfo.getCertSpec().getTlsCert();
+        if (tlsKey == null) {
+            log.warn("tlsKey not present, not providing cert for {}", endpointInfo.getServiceName());
+            return;
+        }
+
+        if (tlsCert == null) {
+            log.warn("tlsCert not present, not providing cert for {}", endpointInfo.getServiceName());
+            return;
+        }
+
+        data.put("tls.key", tlsKey);
+        data.put("tls.crt", tlsCert);
+        log.info("Creating cert secret with certBundle input");
+
+        Secret secret = new SecretBuilder()
+                .editOrNewMetadata()
+                .withName(endpointInfo.getCertSpec().getSecretName())
+                .withNamespace(namespace)
+                .addToLabels(LabelKeys.INFRA_UUID, addressSpace.getAnnotation(AnnotationKeys.INFRA_UUID))
+                .addToLabels(LabelKeys.INFRA_TYPE, addressSpace.getType())
+                .addToLabels("app", "enmasse")
+                .endMetadata()
+                .withType("kubernetes.io/tls")
+                .withData(data)
+                .build();
+
+        Secret existing = client.secrets().inNamespace(namespace).withName(endpointInfo.getCertSpec().getSecretName()).get();
+        if (existing == null) {
+            client.secrets().inNamespace(namespace).createOrReplace(secret);
+        } else if (!existing.getData().equals(secret.getData())) {
```

Comment:
The API does not make is clear if existing.getData(), could return null, but the code would be robust to this possibility if you flipped the equality around.

Reply:
Resolution looks good.